File: 739-daily-temperatures/daily-temperatures.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function(temperatures) {
    // inputs: temperatures = array of numbers
    // outputs: answer[]
    let answer = new Array(temperatures.length).fill(0)
    // O(n^2) 
    let stack = []
    for (let i = 0; i < temperatures.length; i++){
        // always decreasing stack => if we encounter a value that is bigger than
        // most recent stack value we have to keep popping till our stack is decreasing

        while (stack.length > 0 && temperatures[i] > temperatures[stack[stack.length - 1]]) {
            const index = stack.pop()
            answer[index] = i - index
        }

        stack.push(i)
    }
    return answer
};

File: 739-daily-temperatures/daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        # return an array answer
        # answer[i] is the number of days you have to wait after the ith day to get 
        # a warmer temp or 0 if no warmer temp

        answer = [0] * len(temperatures)
        stack = []

        for i,temp in enumerate(temperatures):
            if len(stack) == 0:
                stack.append([i,temp])
                continue
            
            while stack and temp > stack[-1][1]:
                stack_index, stack_temp = stack.pop()
                answer[stack_index] = i - stack_index
            

            stack.append([i, temp])

        return answer